ProjFS.Mac/PrjFSKext/KauthHandler.cpp

```diff
@@ -358,7 +365,8 @@ static int HandleVnodeOperation(
     
     if (isDeleteAction)
     {
-        if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, &root, &vnodeFsidInode, &kauthResult, kauthError))
+        // Allow any user to delete, as this generally doesn't cause nested kauth callbacks.
+        if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, CallbackPolicy_AllowAny, &root, &vnodeFsidInode, &kauthResult, kauthError))
```

Comment:
I don't think we have to make changes now, but I wanted to mention that this code looks a bit strange because we check `CallbackPolicy_AllowAny` here for PreDelete and then below for the PreDelete of directories we use `CallbackPolicy_UserInitiatedOnly`.

```
// Prevent system services from enumerating directories as this tends to cause deadlocks with the kauth listeners for Antivirus software
if (!TryGetVirtualizationRoot(&perfTracer, context, currentVnode, pid, CallbackPolicy_UserInitiatedOnly, &root, &vnodeFsidInode, &kauthResult, kauthError))
{
```

I suspect this can get cleaned up once we have WILL_RENAME as we'll be able to differentiate between PreDelete for rename and PreDelete for delete.

We could consider updating this call to:

```isDirectory ? CallbackPolicy_UserInitiatedOnly : CallbackPolicy_AllowAny```

And including a comment that directory PreDeletes require expansion because they might be part of a rename.

Thoughts?

Reply:
Good point, I've fixed that up and updated the comment to reflect it as well.